src/corefx/System.Globalization.Native/calendarData.cpp

```diff
@@ -0,0 +1,391 @@
+	}
+
+	return false;
+}
+
+/*
+Function:
+InvokeCallbackForDateTimePattern
+
+Gets the DateTime pattern for the specified skeleton and invokes the callback with the retrieved value.
+*/
+bool InvokeCallbackForDateTimePattern(Locale& locale, const char* patternSkeleton, EnumCalendarInfoCallback callback, void* context)
+{
+	UErrorCode err = U_ZERO_ERROR;
+	LocalPointer<DateTimePatternGenerator> generator(DateTimePatternGenerator::createInstance(locale, err));
+	UnicodeString pattern = generator->getBestPattern(UnicodeString(patternSkeleton), err);
+	if (U_SUCCESS(err))
+	{
+		callback(pattern.getTerminatedBuffer(), context);
+		return true;
+	}
+
+	return false;
+}
+
+/*
+Function:
+EnumCalendarArray
+
+Enumerates an array of strings and invokes the callback for each value.
+*/
+bool EnumCalendarArray(const UnicodeString* srcArray, int32_t srcArrayCount, EnumCalendarInfoCallback callback, void* context)
+{
+	bool result = true;
+
+	for (int i = 0; i < srcArrayCount; i++)
+	{
+		UnicodeString src = srcArray[i];
+		result &= callback(src.getTerminatedBuffer(), context);
+	}
+
+	return result;
+}
+
+/*
+Function:
+EnumWeekdays
+
+Enumerates all the weekday names of the specified context and width, invoking the callback function
+for each weekday name.
+*/
+bool EnumWeekdays(
+	DateFormatSymbols& dateFormatSymbols,
+	DateFormatSymbols::DtContextType dtContext,
+	DateFormatSymbols::DtWidthType dtWidth,
+	EnumCalendarInfoCallback callback,
+	void* context)
+{
+	int32_t daysCount;
+	const UnicodeString* dayNames = dateFormatSymbols.getWeekdays(daysCount, dtContext, dtWidth);
+
+	// ICU returns an empty string for the first/zeroth element in the weekdays array.
+	// So skip the first element.
+	dayNames++;
+
+	return EnumCalendarArray(dayNames, daysCount, callback, context);
+}
+
+/*
+Function:
+EnumMonths
+
+Enumerates all the month names of the specified context and width, invoking the callback function
+for each month name.
+*/
+bool EnumMonths(
+	DateFormatSymbols& dateFormatSymbols,
+	DateFormatSymbols::DtContextType dtContext,
+	DateFormatSymbols::DtWidthType dtWidth,
+	EnumCalendarInfoCallback callback,
+	void* context)
+{
+	int32_t monthsCount;
+	const UnicodeString* monthNames = dateFormatSymbols.getMonths(monthsCount, dtContext, dtWidth);
+	return EnumCalendarArray(monthNames, monthsCount, callback, context);
+}
+
+/*
+Function:
+EnumCalendarInfo
+
+Retrieves a collection of calendar string data specified by the locale, calendar, and data type.
+Allows for a collection of calendar string data to be retrieved by invoking
+the callback for each value in the collection.
+The context parameter is passed through to the callback along with each string.
+*/
+extern "C" bool EnumCalendarInfo(
+	EnumCalendarInfoCallback callback,
+	UChar* localeName,
+	CalendarId calendarId,
+	CalenderDataType dataType,
+	void* context)
+{
+	UErrorCode err = U_ZERO_ERROR;
+
+	Locale locale = GetLocale(localeName);
+	if (locale.isBogus())
```

Comment:
FWIW I prefer switch() here. One benefit is you will get a compiler warning if not all cases are accounted for.